src/redisearch_rs/inverted_index/src/full.rs

```diff
@@ -0,0 +1,194 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+
+use std::{
+    io::{Cursor, Seek, Write},
+    mem::ManuallyDrop,
+};
+
+use ffi::{t_docId, t_fieldMask};
+use qint::{qint_decode, qint_encode};
+use varint::VarintEncode;
+
+use crate::{Decoder, Encoder, RSIndexResult, RSOffsetVector, RSResultType, RSTermRecord};
+
+/// Encode and decode the delta, frequency, field mask and offsets of a term record.
+///
+/// This encoder supports field masks fitting in a `u32`.
+/// Use [`FullWide`] for `u128` field masks.
+///
+/// The delta, frequency, field mask and offsets lengths are encoded using [qint encoding](qint).
+///
+/// The offsets themselves are then written directly.
+///
+/// This encoder only supports delta values that fit in a `u32`.
+#[derive(Default)]
+pub struct Full;
+
+/// Return a slice of the offsets vector from a term record.
+///
+/// # Safety
+///
+/// record must have `result_type` set to `RSResultType::Term`.
+#[inline(always)]
+fn offsets<'a>(record: &'a RSIndexResult<'_>) -> &'a [u8] {
+    // SAFETY: caller ensured the proper result_type.
+    let term = record.as_term().unwrap();
+    if term.offsets.data.is_null() {
+        &[]
+    } else {
+        assert_eq!(term.offsets.len, record.offsets_sz);
+
+        // SAFETY: We checked that data is not NULL and `len` is guaranteed to be a valid length for the data pointer.
+        unsafe {
+            std::slice::from_raw_parts(term.offsets.data as *const u8, term.offsets.len as usize)
+        }
+    }
+}
+
+impl Encoder for Full {
+    type Delta = u32;
+
+    fn encode<W: Write + Seek>(
+        &mut self,
+        mut writer: W,
+        delta: Self::Delta,
+        record: &RSIndexResult,
+    ) -> std::io::Result<usize> {
+        assert!(matches!(record.result_type, RSResultType::Term));
+
+        let field_mask = record
+                .field_mask
+                .try_into()
+                .expect("Need to use the wide variant of the Full encoder to support field masks bigger than u32");
+
+        let mut bytes_written = qint_encode(
+            &mut writer,
+            [delta, record.freq, field_mask, record.offsets_sz],
+        )?;
+
+        let offsets = offsets(record);
+        bytes_written += writer.write(offsets)?;
+
+        Ok(bytes_written)
+    }
+}
+
+/// Create a [`RSIndexResult`] from the given parameters and read its offsets from the reader.
+#[inline(always)]
+fn decode_term_record_offsets<'a>(
+    cursor: &mut Cursor<&'a [u8]>,
+    base: t_docId,
+    delta: u32,
+    field_mask: t_fieldMask,
+    freq: u32,
+    offsets_sz: u32,
+) -> std::io::Result<RSIndexResult<'a>> {
```

Comment:
Is there a performance hit when the `inline` is missing?

Reply:
I removed it.

Reply:
The performance hit seems to be bigger with the other offsets decoder using this function, so I'd rather keep it if that's ok with you?